### ISSUE TITLE
[HOLD for payment 2024-02-07] MEDIUM: [$500] Drop domain name in mentions if chat members are on the same domain

### ISSUE THREAD

**Issue by NikkiWines** [Expensify/App#32534]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.8-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A 
**Email or phone of affected tester (no customers):** all customers on private domains
**Logs:** N/A 
**Expensify/Expensify Issue URL:** N/A
**Issue reported by:** N/A
**Slack conversation:** https://expensify.slack.com/archives/C066HJM2CAZ/p1701743829149999

## Action Performed:
1. Create a room with 4 members: 2 on the same private domain, 1 on another private domain, and 1 on a public domain 
2. As each member, send a message mentioning each other member 

## Expected Result:
If the user looking at the chat is on the same domain as the person being mentioned, they should not see the domain (e.g. `@user@expensify.com`) in the mention. They should see a shortened mention containing only the email prefix (e.g. `@user`) 

If the user looking at the chat is not on the same domain as the person being mentioned, or if the user being mentioned is on a public domain, then they should see the full mention (e.g. `@user@gmail.com`) 

E.g. viewing as `nikki@meep.com`, which has a common domain
![image](https://github.com/Expensify/App/assets/16747822/b79caef7-23eb-47a7-948d-d3d1ea46576a)

E.g. viewing as `nikki@notification.com`, no common domain
![image](https://github.com/Expensify/App/assets/16747822/3648c222-a04c-425e-abd7-868f05a725ef)

## Actual Result:
Regardless of whether or not users share a domain, the mention is always the full mention (e.g. `@user@expensify.com`) 

E.g. viewing as `nikki@meep.com`, which has a common domain
![image](https://github.com/Expensify/App/assets/16747822/b79caef7-23eb-47a7-948d-d3d1ea46576a)

E.g. viewing as `nikki@notification.com`, no common domain
![image](https://github.com/Expensify/App/assets/16747822/3364cbc3-596e-43f0-afbc-814dd0e00907)

## Workaround:
N/A

## Platforms:
Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

N/A 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0173ab1d6ef6e3a31a</li>
        <li>Upwork Job ID: 1732819861042143232</li>
        <li>Last Price Increase: 2023-12-29</li>
    </ul>
</details>

**Comment by puneetlath** [comment]
A few things to add here: 

- This should only be the case for private domains (e.g. expensify.com). For public domains (e.g. gmail.com) should always use the full email. We have a list of public domains in expensify-common [here](https://github.com/Expensify/expensify-common/blob/073ec36d75d7b214e6d0220d35e600bdb0a9e97f/lib/CONST.d.ts#L737). So if my email address is `puneet@gmail.com` and I'm viewing a mention of `nikki@gmail.com` it should show as `@nikki@gmail.com` not `@nikki`
- The app supports both email based mentions - `<mention-user>@person@gmail.com</mention-user>` - and accountID-based mentions - `<mention-user accountID=1234></mention-here>`. We should make sure that the solution works for both of them. So I would think that maybe it's as follows. 

For email-based mentions: 
1. Look at the email address in the mention
2. If it's on the same private domain as mine, render as `@person`. Otherwise render as `@person@domain.com`

For accountID-based mentions:
1. Look at the personalDetails object for the accountID
2. If we don't have a personalDetail for that accountID, render as `@Hidden`
3. If we have a personalDetail, but we don't have a display name or login in there, render as `@Hidden`
4. If we have a personalDetail with display name, but no login, render as `@Display Name`
5. if we have a personalDetail with login, render the same way as email-based mentions

Does that sound good?

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
MEDIUM: Drop domain name in mentions if chat members are on the same domain
### What is the root cause of that problem?
NA - Feature Request
### What changes do you think we should make in order to solve the problem?

We just need to render through the rules of things first flow.

1. Checking if the rendering is mention is a email or not.
```jsx
// Checking if the rendering mention is login i.e email
    const isDisplyingLogin = displayNameOrLogin === lodashGet(user, 'login', '');
    // Check if the email is valid, if so extract the email
    const mentionUserEmailDomain = isDisplyingLogin && Str.isValidEmail(displayNameOrLogin) && Str.extractEmailDomain(displayNameOrLogin);
```
2. Extracting domains form the emails

```jsx
    const currentUserEmailDomian = Str.extractEmailDomain(props.currentUserPersonalDetails.login);
```

3. If the the domain is from public domains
    const isPublicDomain = StrConst.PUBLIC_DOMAINS.includes(mentionUserEmailDomain.toLowerCase());

4. if the both users (mentioned and viewing) are having same domains
 const isSameOriginEmail = mentionUserEmailDomain === currentUserEmailDomian;

5. Update the rendering value.
```jsx
if (isSameOriginEmail && isDisplyingLogin && !isOurMention && !isPublicDomain) {
        displayNameOrLogin = Str.replaceAll(`@${currentUserEmailDomian}`, '');
    }


```


### What alternative solutions did you explore? (Optional)
NA

**Comment by quinthar** [comment]
FYI, before doing anything on this, please read this [entire @mentions design doc](https://docs.google.com/document/d/1vhvpwQDStA0lcsI0Z33pXOKPM6nXGRuplvQTFBWvCyA/edit#heading=h.ns12dr1f35rx) to make sure you understand the full context.  Also, be sure to call out anything inaccurate.  If we're going to do something different than the doc says, or do it in a different order, let's update the doc FIRST to reflect the real plan -- and THEN do the coding.  Thanks!

**Comment by b4s36t4** [comment]
@quinthar design doc is protected, can you please share the link with read permission? thanks

**Comment by quinthar** [comment]
Ah, what's your email address so I can share it with you?

**Comment by b4s36t4** [comment]
@quinthar it's `maheshvagicherla99438@gmail.com`

**Comment by quinthar** [comment]
Shared!

**Comment by b4s36t4** [comment]
Got it, thanks :)

**Comment by quinthar** [comment]
Hm, reading the design doc, and I'm not actually sure it ever finished the design for "simplified handles".  @NikkiWines are you sure this is ready for a contributor to pick up?  I think we should first finish the design doc for this to make sure we all agree on how it's going to work, before taking it out for proposals.

**Comment by b4s36t4** [comment]
<img width="725" alt="Screenshot 2023-12-08 at 1 45 57 PM" src="https://github.com/Expensify/App/assets/59088937/2f437894-1acd-4a0e-80a2-4772e27b8733">

Wasn't this the specification?

**Comment by quinthar** [comment]
@b4s36t4 - I agree that provides some very high level of detail, but I think there are a lot of outstanding details.  In particular, how does the client know if the domain is "public"?  We have that on the server, but I don't see how the client would know.  It also isn't clear to me how we actually dynamically alter the mention at the time of rendering.

**Comment by quinthar** [comment]
@szebniok - It looks like you did a lot of work on mentions, what's your take on how we can render "simplified mentions" when on the same domain as the person being mentioned?  Are there any tricky gotchas we should be aware of?

**Comment by b4s36t4** [comment]
>  It also isn't clear to me how we actually dynamically alter the mention at the time of rendering

Answering this, we render mention based on 3 things i.e `login`, `displayName` and `accountID` param from the html. At this point I don't see any issue with `displayName`

More on it here https://github.com/Expensify/App/blob/226f885ddee6033bbce5ae182f5a74586ee4a5c0/src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js#L45

Since we only have to simplify mentions for emails i.e we need to depend on `login` and `accountID` where the mention logic is depends on these two values.

Checks we should have when rendering.

* We're rendering the user's login & that is too email.
* If the email is valid (above one should handle that)
* Domain handling, we have the domain validation, email domain stripping everything inside `expensify-common` lib.
* User's domain verification, get the domain of the current logged in user (if the mention is not `@here` or so)
* check for similarity of email domain, even if they match it shouldn't be from a public domain.

Also one doubt I have is say I'm from `mahesh@demo.com` what should we render for the same user, `mahesh` or `mahesh@demo.com` (not worrying about public domain though).

**Comment by stephanieelliott** [comment]
@sobitneupane @tienifr I've shared the doc with you!

**Comment by NikkiWines** [comment]
Dropping this down to weekly while it's on hold

**Comment by NikkiWines** [comment]
on hold, not overdue.

**Comment by quinthar** [comment]
Hey all, now that I've caught up with the doc and this discussion... it feels like you all know what's going on way better than me!  I'm taking this off hold.

@sobitneupane what are the next steps -- do you evaluate and select the proposals?  Let's get this party started!

**Comment by quinthar** [comment]
Who are we waiting on to review the existing proposals?  Is that @NikkiWines or @sobitneupane or @stephanieelliott ?  What are the next steps?

**Comment by sobitneupane** [comment]
Thanks for the proposals.

@puneetlath has already [briefed](https://github.com/Expensify/App/issues/32534#issuecomment-1843288493) about the expected outcome and has outlined the solution. 

@tienifr Can you please include some implementation details in your solution?

@b4s36t4 Where will you make the proposed changes? Does it cover account-ID based mentions as well? Even account-ID based mention might have login with the same domain, if I am not wrong.

Will the domain name in mentions dropped from the MentionSuggestions list as well? Or will the change reflect in report only?

**Comment by b4s36t4** [comment]
@sobitneupane the changes are done in `MentionUser.js` which is the custom Component we use to render the `mention` tag.

The solution I explained is straight solution, first we solve the rendering value (based on email or accountID the existing value never changes), once we have the current rendering value then we do the new changes i.e checking for email.

**Comment by sobitneupane** [comment]
Thanks for the proposal everyone.

Implementation approach is similar in all the proposals.

https://github.com/Expensify/App/blob/789c09491839f23f4a86a321bc7dfdffb675542d/src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.js#L86

The above line is the one responsible for displaying user-mentions. If accountId is not present in the node, we use the second expression to display user-mention. @b4s36t4 Your proposal hasn't discussed about the 2nd case.

**Comment by sobitneupane** [comment]
@gijoe0295 Thanks for your proposal. Your proposal in similar to the previous proposals.

**Comment by sobitneupane** [comment]
Thanks for the updated @tienifr 

[Proposal](https://github.com/Expensify/App/issues/32534#issuecomment-1845888229) from @tienifr looks good to me.

🎀 👀 🎀 C+ reviewed

**Comment by NikkiWines** [comment]
Yep, agreed! @tienifr's proposal looks good!

**Comment by NikkiWines** [comment]
Oh wait, gotta apply labels and such - one moment

**Comment by quinthar** [comment]
Great, what's the next step for accepting @tienifr's proposal and getting started?

**Comment by sobitneupane** [comment]
@tienifr has created [PR](https://github.com/Expensify/App/pull/34013), and it is currently under review.

**Comment by stephanieelliott** [comment]
PR is still being actively reviewed!

**Comment by NikkiWines** [comment]
🤔 good question - I would say yes, there's not really a reason you'd need to see your own domain name

**Comment by puneetlath** [comment]
I think we should treat your own domain same as any other (so don't cut it down if it's public domain).

**Comment by NikkiWines** [comment]
Oh yeah, good distinction - i was only thinking of the private domain case 😅

**Comment by stephanieelliott** [comment]
Hey @sobitneupane looks the PR is waiting your feedback, will you be able to resume review soon? https://github.com/Expensify/App/pull/34013#issuecomment-1888423660

**Comment by quinthar** [comment]
@sobitneupane what's your ETA for feedback and review?  This is a super amazing feature, let's get it out!

**Comment by sobitneupane** [comment]
There were some issues in the PR. I am waiting @tienifr to resolve those issues.

**Comment by quinthar** [comment]
@tienifr what is your ETA?  I'm extremely eager to get this out.

**Comment by quinthar** [comment]
Great!  Sobitreupane, can you please respond immediately?  Thanks!

**Comment by sobitneupane** [comment]
I am reviewing the PR now.

**Comment by stephanieelliott** [comment]
C+ review has been completed, PR is now just waiting on a few minor change requests to be addressed by @tienifr before merge.

**Comment by stephanieelliott** [comment]
PR is updated, just awaiting a final review

**Comment by stephanieelliott** [comment]
PR is merged and undergoing QA

**Comment by stephanieelliott** [comment]
This PR is on staging 👏

**Comment by stephanieelliott** [comment]
Hey @sobitneupane can you complete the BZ checklist when you get a chance?

**Comment by stephanieelliott** [comment]
Bump @sobitneupane, can you complete the BZ checklist?

**Comment by sobitneupane** [comment]
It's a feature request not a bug. So, I don't think the checklist applies here.

**Comment by stephanieelliott** [comment]
Testrail issue created: https://github.com/Expensify/Expensify/issues/368461

@tienifr offer extended in Upwork, please accept when you get a chance!

**Comment by stephanieelliott** [comment]
All paid, thanks!

**Comment by JmillsExpensify** [comment]
$500 approved for @sobitneupane based on [this summary](https://github.com/Expensify/App/issues/32534#issuecomment-1932246849).
